File: VB/DXGridDataPager/SourcesBehavior.vb
Imports Microsoft.VisualBasic
Imports System.Windows.Interactivity
Imports DevExpress.Xpf.Editors.DataPager
Imports System.Windows
Imports System.Collections
Imports System.Collections.Specialized
Namespace DXGridDataPager
	Public Class SourcesBehavior
		Inherits Behavior(Of DataPager)
		Public Shared ReadOnly ActualSourceProperty As DependencyProperty = DependencyProperty.Register("ActualSource", GetType(Object), GetType(SourcesBehavior), Nothing)
		Public Shared ReadOnly SourcesProperty As DependencyProperty = DependencyProperty.Register("Sources", GetType(IList), GetType(SourcesBehavior), New PropertyMetadata(Nothing, Function(d, e) (CType(d, SourcesBehavior)).OnSourcesChanged(CType(e.NewValue, IList))))
		Public Property ActualSource() As Object
			Get
				Return CObj(GetValue(ActualSourceProperty))
			End Get
			Set(ByVal value As Object)
				SetValue(ActualSourceProperty, value)
			End Set
		End Property
		Public Property Sources() As IList
			Get
				Return CType(GetValue(SourcesProperty), IList)
			End Get
			Set(ByVal value As IList)
				SetValue(SourcesProperty, value)
			End Set
		End Property
		Public ReadOnly Property DataPager() As DataPager
			Get
				Return AssociatedObject
			End Get
		End Property
		Protected Overrides Sub OnAttached()
			MyBase.OnAttached()
			'DataPager.PageSize = Sources.Count;
			DataPager.ItemCount = Sources.Count
			'DataPager.PageSize = Sou
			UpdateActualSource(DataPager.PageIndex)
			SubsribeSourcesColletion(Sources)
			SubscribeDataPager()
		End Sub
		Protected Overrides Sub OnDetaching()
			UnsubsribeDataPager()
			UnsubsribeSourcesColletion(Sources)
			MyBase.OnDetaching()
		End Sub
		Private Function OnSourcesChanged(ByVal oldSources As IList) As Object
			If DataPager Is Nothing Then
				Return Nothing
			End If
			UpdateActualSource(DataPager.PageIndex)
			If Sources IsNot Nothing Then
				DataPager.ItemCount = Sources.Count
			End If
			UnsubsribeSourcesColletion(oldSources)
			SubsribeSourcesColletion(Sources)
			If ActualSource IsNot Nothing Then
				DataPager.PageIndex = Sources.IndexOf(ActualSource)
			End If
			SubscribeDataPager()
			Return Nothing
		End Function
		Private Sub UnsubsribeSourcesColletion(ByVal sources As IList)
			If TypeOf sources Is INotifyCollectionChanged Then
				RemoveHandler (CType(sources, INotifyCollectionChanged)).CollectionChanged, AddressOf Sources_CollectionChanged
			End If
		End Sub
		Private Sub SubsribeSourcesColletion(ByVal sources As IList)
			UnsubsribeSourcesColletion(sources)
			If TypeOf sources Is INotifyCollectionChanged Then
				AddHandler (CType(sources, INotifyCollectionChanged)).CollectionChanged, AddressOf Sources_CollectionChanged
			End If
		End Sub
		Private Sub UnsubsribeDataPager()
			If DataPager Is Nothing Then
				Return
			End If
			RemoveHandler DataPager.PageIndexChanged, AddressOf DataPager_PageIndexChanged
		End Sub
		Private Sub SubscribeDataPager()
			If DataPager Is Nothing Then
				Return
			End If
			UnsubsribeDataPager()
			AddHandler DataPager.PageIndexChanged, AddressOf DataPager_PageIndexChanged
		End Sub

		Private Sub Sources_CollectionChanged(ByVal sender As Object, ByVal e As System.Collections.Specialized.NotifyCollectionChangedEventArgs)
			If Sources IsNot Nothing Then
				DataPager.ItemCount = Sources.Count
			End If
			If ActualSource Is Nothing Then
				UpdateActualSource(DataPager.PageIndex)
			End If
		End Sub
		Private Sub UpdateActualSource(ByVal index As Integer)
			If Sources IsNot Nothing AndAlso Sources.Count > 0 Then
				If index < Sources.Count Then
					ActualSource = Sources(index)
				Else
					ActualSource = Sources(0)
				End If
			End If

		End Sub
		Private Sub DataPager_PageIndexChanged(ByVal sender As Object, ByVal e As DataPagerPageIndexChangedEventArgs)
			UpdateActualSource(e.NewValue)
		End Sub
	End Class
End Namespace

File: CS/DXGridDataPager/SourcesBehavior.cs
using System.Windows.Interactivity;
using DevExpress.Xpf.Editors.DataPager;
using System.Windows;
using System.Collections;
using System.Collections.Specialized;
namespace DXGridDataPager {
    public class SourcesBehavior : Behavior<DataPager> {
        public static readonly DependencyProperty ActualSourceProperty =
            DependencyProperty.Register("ActualSource", typeof(object), typeof(SourcesBehavior), null);
        public static readonly DependencyProperty SourcesProperty =
            DependencyProperty.Register("Sources", typeof(IList), typeof(SourcesBehavior),
            new PropertyMetadata(null, (d, e) => ((SourcesBehavior)d).OnSourcesChanged((IList)e.NewValue)));
        public object ActualSource {
            get { return (object)GetValue(ActualSourceProperty); }
            set { SetValue(ActualSourceProperty, value); }
        }
        public IList Sources {
            get { return (IList)GetValue(SourcesProperty); }
            set { SetValue(SourcesProperty, value); }
        }
        public DataPager DataPager { get { return AssociatedObject; } }
        protected override void OnAttached() {
            base.OnAttached();
            //DataPager.PageSize = Sources.Count;
            DataPager.ItemCount = Sources.Count;
            //DataPager.PageSize = Sou
            UpdateActualSource(DataPager.PageIndex);
            SubsribeSourcesColletion(Sources);
            SubscribeDataPager();
        }
        protected override void OnDetaching() {
            UnsubsribeDataPager();
            UnsubsribeSourcesColletion(Sources);
            base.OnDetaching();
        }
        void OnSourcesChanged(IList oldSources) {
            if(DataPager == null) return;
            UpdateActualSource(DataPager.PageIndex);
            if(Sources != null) DataPager.ItemCount = Sources.Count;
            UnsubsribeSourcesColletion(oldSources);
            SubsribeSourcesColletion(Sources);
            if(ActualSource != null) DataPager.PageIndex = Sources.IndexOf(ActualSource);
            SubscribeDataPager();
        }
        void UnsubsribeSourcesColletion(IList sources) {
            if(sources is INotifyCollectionChanged)
                ((INotifyCollectionChanged)sources).CollectionChanged -= Sources_CollectionChanged;
        }
        void SubsribeSourcesColletion(IList sources) {
            UnsubsribeSourcesColletion(sources);
            if(sources is INotifyCollectionChanged)
                ((INotifyCollectionChanged)sources).CollectionChanged += Sources_CollectionChanged;
        }
        void UnsubsribeDataPager() {
            if(DataPager == null) return;
            DataPager.PageIndexChanged -= DataPager_PageIndexChanged;
        }
        void SubscribeDataPager() {
            if(DataPager == null) return;
            UnsubsribeDataPager();
            DataPager.PageIndexChanged += DataPager_PageIndexChanged;
        }

        void Sources_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) {
            if(Sources != null) DataPager.ItemCount = Sources.Count;
            if(ActualSource == null) UpdateActualSource(DataPager.PageIndex);
        }
        void UpdateActualSource(int index) {
            if(Sources != null && Sources.Count > 0) {
                if(index < Sources.Count) ActualSource = Sources[index];
                else ActualSource = Sources[0];
            }

        }
        void DataPager_PageIndexChanged(object sender, DataPagerPageIndexChangedEventArgs e) {
            UpdateActualSource(e.NewValue);
        }
    }
}